src/silentpayments.rs

```diff
@@ -0,0 +1,546 @@
+            label_tweak32.as_mut_c_ptr(),
+            recipient_scan_key.as_c_ptr(),
+            m,
+        );
+
+        if res == 1 {
+            let pubkey = PublicKey::from(pubkey);
+            let label_tweak = label_tweak32;
+
+            return Ok(LabelTweakResult { pubkey, label_tweak });
+        } else {
+            return Err(LabelTweakError::Failure);
+        }
+    }
+}
+
+/// Struct to store public data
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct SilentpaymentsPublicData(ffi::SilentpaymentsPublicData);
+
+impl CPtr for SilentpaymentsPublicData {
+    type Target = ffi::SilentpaymentsPublicData;
+
+    /// Obtains a const pointer suitable for use with FFI functions.
+    fn as_c_ptr(&self) -> *const Self::Target { &self.0 }
+
+    /// Obtains a mutable pointer suitable for use with FFI functions.
+    fn as_mut_c_ptr(&mut self) -> *mut Self::Target { &mut self.0 }
+}
+
+/// Label tweak errors
+#[derive(Debug, Clone, Copy, Eq, PartialEq, PartialOrd, Ord, Hash)]
+pub enum SilentpaymentsPublicDataError {
+    /// Failed to create the public data
+    CreationFailure,
+    /// Serialization Failure
+    SerializationFailure,
+    /// Parse Failure
+    ParseFailure,
+    /// Failed to create the shared secret
+    SharedSecretFailure
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for SilentpaymentsPublicDataError {}
+
+impl fmt::Display for SilentpaymentsPublicDataError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+        match self {
+            SilentpaymentsPublicDataError::SerializationFailure => write!(f, "Failed to serialize silent payments public data"),
+            SilentpaymentsPublicDataError::ParseFailure => write!(f, "Failed to parse silent payments public data"),
+            SilentpaymentsPublicDataError::SharedSecretFailure => write!(f, "Failed to create shared secret"),
+            SilentpaymentsPublicDataError::CreationFailure => write!(f, "Failed to create silent payments public data"),
+        }
+    }
+}
+
+impl SilentpaymentsPublicData {
+    /// Create a new `SilentpaymentsPublicData` object.
+    pub fn new() -> SilentpaymentsPublicData {
+        let empty_data = [0u8; constants::SILENT_PAYMENTS_PUBLIC_DATA_SIZE];
+        SilentpaymentsPublicData(ffi::SilentpaymentsPublicData::from_array(empty_data))
+    }
+
+    /// Compute Silent Payment public data from input public keys and transaction inputs
+    pub fn create<C: Verification>(
+        secp: &Secp256k1<C>,
+        smallest_outpoint: &[u8; 36],
+        xonly_pubkeys: Option<&[&XOnlyPublicKey]>,
+        plain_pubkeys: Option<&[&PublicKey]>,
+    ) -> Result<Self, SilentpaymentsPublicDataError> {
+
+        let cx = secp.ctx().as_ptr();
+
+        unsafe {
+            
+            let empty_data = [0u8; constants::SILENT_PAYMENTS_PUBLIC_DATA_SIZE];
+            let mut silentpayments_public_data = ffi::SilentpaymentsPublicData::from_array(empty_data);
```

Comment:
This can use `Self::new()`, and `as_mut_c_ptr()` for the c function. It then does not need to be wrapped before returning.

Same comment for line 308.

Reply:
Thanks. Done in 6fbb0595e36f9acf4f24e20e43c7c790ee91f22e